src/evict/evict_lru.c

```diff
@@ -1637,14 +1659,9 @@ __evict_walk(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue)
     }
 
 err:
-    if (dhandle_locked)
+    if (dhandle_list_locked)
         __wt_readunlock(session, &conn->dhandle_lock);
 
```

Comment:
This cleanup logic is now a part of "set" and "clear" functions.

This function intends to keep `cache->walk_tree` across calls so that next time it's called, it can proceed to the next dhandle in list.

Decrementing `dhandle->session_inuse` here is an error since the dhandle is still stored in `cache->walk_tree` (but its use count is decremented). It can potentially be freed by sweep server.

Reply:
We discussed this offline. The existing clean up us incorrect as it decrements session_inuse, but doesn't unassign the dhandle. This used to leave us open to holding a pointer to a dhandle which could be cleared by the sweep server.

The new logic pairs decrementing of session_inuse with the unassignment of the dhandle, which is the correct behaviour.